draft-ietf-quic-recovery.md

```diff
@@ -337,29 +337,44 @@ samples, and rttvar is the variation in the RTT samples, estimated using a
 mean variation.
 
 The calculation of smoothed_rtt uses path latency after adjusting RTT samples
-for acknowledgement delays. These delays are computed using the ACK Delay field
-of the ACK frame as described in Section 19.3 of {{QUIC-TRANSPORT}}. For packets
-sent in the Application Data packet number space, a peer limits any delay in
-sending an acknowledgement for an ack-eliciting packet to no greater than the
-value it advertised in the max_ack_delay transport parameter. Consequently, when
-a peer reports an acknowledgment delay that is greater than its max_ack_delay,
-the delay is attributed to reasons out of the peer's control, such as scheduler
-latency at the peer or loss of previous ACK frames.  Any delays beyond the
-peer's max_ack_delay are therefore considered effectively part of path delay and
-incorporated into the smoothed_rtt estimate.
+for acknowledgement delays. These delays are computed using the ACK Delay
+field of the ACK frame as described in Section 19.3 of {{QUIC-TRANSPORT}}.
+
+When 0-RTT or 1-RTT ack-eliciting packets are received, a peer MUST NOT delay
+acknowledging them any longer than the period it advertised in the max_ack_delay
+transport parameter (Section 18.2 of {{QUIC-TRANSPORT}}), with the following
+exception.  Prior to handshake confirmation, the peer might not have the packet
+protection keys for these packets when they are received. It might therefore
+buffer them and acknowledge them when the requisite keys become available.
+
+Since the peer might report large acknowledgement delays, the endpoint MAY
+ignore max_ack_delay until the handshake is confirmed (Section 4.1.2 of
+{{QUIC-TLS}}). Since these acknowledgement delays, when they occur, are
+measurable and limited to the handshake, the endpoint can use them without
+limiting them to the max_ack_delay and avoid unnecessarily inflating the
+smoothed_rtt estimate.
+
+After the handshake is confirmed, any acknowledgement delays reported by the
+peer that are greater than its max_ack_delay are attributed to unintentional but
+potentially repeating delays, such as scheduler latency at the peer or loss of
+previous acknowledgements. Any delays beyond the peer's max_ack_delay are
+therefore considered effectively part of path delay and incorporated into the
+smoothed_rtt estimate.
 
 When adjusting an RTT sample using peer-reported acknowledgement delays, an
 endpoint:
 
-- MUST ignore the ACK Delay field of the ACK frame for packets sent in the
-  Initial and Handshake packet number space.
+- MAY ignore the ACK Delay field of the ACK frame for packets sent in the
+  Initial packet number space.
 
-- MUST use the lesser of the value reported in ACK Delay field of the ACK frame
-  and the peer's max_ack_delay transport parameter.
+- MAY ignore the peer's max_ack_delay until the handshake is confirmed,
```

Comment:
Do we want to say SHOULD here instead, given that buffering happens?

It means updating pseudocode, but the change is small.

Reply:
I think SHOULD is good.

Reply:
@ianswett may have an opinion here. I'll wait for it.

Reply:
+1 to changing it to SHOULD. As others have pointed out, there are cases that the delay is inevitable (e.g., due to a client doing OCSP cert validation).

Reply:
I think that's ok, but is there a reason line 371 is a MAY instead of a SHOULD?  I can't think of any.  If it was a SHOULD, the the pseudocode should follow the SHOULDs.

Reply:
MAY is in recognition of the fact that an attacker can already tilt the calculation for RTT using Initial ACK, so we allow people to ignore that information, but we don't need to recommend it.